File: php/src/Advertisement/Domain/Model/ValueObject/AdvertisementId.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisement\Domain\Model\ValueObject;

use Chemem\Bingo\Functional\Functors\Monads\Either;
use Demo\App\Advertisement\Domain\Exceptions\InvalidUniqueIdentifierException;

final readonly class AdvertisementId
{
    private function __construct(private string $value) {}

    public static function build(string $value): Either
    {
        if (!self::validate($value)) {
            return Either::left(InvalidUniqueIdentifierException::withId($value));
        }

        return Either::right(new self($value));
    }

    public function value(): string
    {
        return $this->value;
    }

    private static function validate(string $value): bool
    {
        return preg_match('/^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i', $value) != 0;
    }
}

File: kotlin/src/main/kotlin/advertisement/domain/model/value_object/AdvertisementId.kt
package advertisement.domain.model.value_object

import advertisement.domain.errors.InvalidAdvertisementIdFormatError
import arrow.core.Either
import arrow.core.left
import arrow.core.right

class AdvertisementId private constructor(private val value: String) {

    companion object {
        private val regex = "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-4[0-9a-fA-F]{3}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$".toRegex()

        fun build(value: String): Either<InvalidAdvertisementIdFormatError, AdvertisementId> =
            when {
                !value.matches(regex) -> InvalidAdvertisementIdFormatError.withId(value).left()
                else -> AdvertisementId(value).right()
            }
    }

    fun value(): String = value
}
